lib/Immediate/Immediate.cpp

```diff
@@ -432,3 +434,59 @@ int swift::RunImmediatelyFromAST(CompilerInstance &CI) {
 
   return *Result;
 }
+
+void swift::execWithFoundationIfNotLoaded(ArrayRef<const char *> Args,
+                                          const char *Argv0,
+                                          DiagnosticEngine &Diags) {
+#if defined(__APPLE__)
+  const char *FoundationPath =
+      "/System/Library/Frameworks/Foundation.framework/Foundation";
+  void *handle = dlopen(FoundationPath, RTLD_NOLOAD);
+  if (handle) {
+    // Foundation is already loaded. Use dlclose to release the ref-count that
+    // was incremented by dlopen and return.
+    dlclose(handle);
+    return;
+  }
+
+  SmallString<70> EnvStorage("DYLD_INSERT_LIBRARIES=");
+  SmallVector<const char *> NewEnv;
+  for (const char **envIt = environ; *envIt != nullptr; ++envIt) {
+    StringRef EnvVar(*envIt);
+    if (EnvVar.starts_with(EnvStorage)) {
+      if (EnvVar.contains(FoundationPath)) {
+        // Already tried to load Foundation but something went wrong.
+        Diags.diagnose(SourceLoc(),
+                       diag::warning_immediate_mode_cannot_load_foundation,
+                       "library not found in process");
+        return;
+      }
+      EnvStorage.assign(EnvVar);
+      continue;
+    }
+    NewEnv.push_back(*envIt);
+  }
+
+  if (EnvStorage.back() != '=')
```

Comment:
Even more paranoid and contrived than my first comment: this will be a false positive if a value is set and the last path ends in a `=`.